### PR TITLE
BUG: compression_matrix fixed promoting all matrices to 64 bits #2849

### PR DESCRIPTION
**Pull request by RogerMoens**
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

This prevents from promoting datatypes of less then 4 bytes precision to promote to 8 bytes as explained in #2849. This also improves the performances drastically as the following QR and SVD will be faster.

I am working on some extra tests for the randomized svd in general. 

Note: also see #6757

### CODE REVIEW

**Comment by RogerMoens** [comment]
A minimal test:

```python
import numpy as np
import dask.array as da
import time as time
import dask

# Compare dtypes
a = da.random.random(size=(10000,100))
u, s, vt = da.linalg.svd_compressed(a, k=5)
print('a.dtype:', a.dtype, 'u.dtype:', u.dtype, 's.dtype:', s.dtype, 'vt.dtype:', vt.dtype)

b = da.random.random(size=(10000,100)).astype(np.float32, copy=False)
q, r, zt = da.linalg.svd_compressed(b, k=5)
print('b.dtype:', b.dtype, 'q.dtype:', q.dtype, 'r.dtype:', r.dtype, 'zt.dtype:', zt.dtype)

# Compare Calculation Time
# float64
tic = time.time()
u, s, vt = dask.compute(u, s, vt)
print('float64:', time.time()-tic)

# float32
tic = time.time()
q, r, zt = dask.compute(q, r, zt)
print('float32:', time.time()-tic)
```

returns:

```
a.dtype: float64 u.dtype: float64 s.dtype: float64 vt.dtype: float64
b.dtype: float32 q.dtype: float32 r.dtype: float32 zt.dtype: float32
float64: 0.21895766258239746
float32: 0.0695192813873291
```

**Comment by RogerMoens** [comment]
A minimal test for accuracy comparison:

```python
import numpy as np
import dask.array as da
import dask
import scipy.linalg as sl

# Compare dtypes
a = da.random.random(size=(100000, 100))
b = a.astype(np.float32, copy=True)
n_sv = 10

u, s, vt = da.linalg.svd_compressed(a, k=2*n_sv, n_power_iter=2) # Float64
q, r, zt = da.linalg.svd_compressed(b, k=2*n_sv, n_power_iter=2) # Float32
m, l, nt = sl.svd(a, full_matrices=False) # True SVD: GESDD

recon_true = (m[:,:n_sv]*s[:n_sv]).dot(nt[:n_sv,:])
recon_a = (u[:,:n_sv]*s[:n_sv]).dot(vt[:n_sv,:])
recon_b = (q[:,:n_sv]*r[:n_sv]).dot(zt[:n_sv,:])

score = 100*da.linalg.norm(recon_a-recon_b, 'fro')/da.linalg.norm(recon_true, 'fro')
print('Relative Frobenius Difference: ', score.compute(), '%')

score = 100*da.linalg.norm(recon_a-recon_true, 'fro')/da.linalg.norm(recon_true, 'fro')
print('Relative Frobenius Difference Float64: ', score.compute(), '%')

score = 100*da.linalg.norm(recon_b-recon_true, 'fro')/da.linalg.norm(recon_true, 'fro')
print('Relative Frobenius Difference Float32: ', score.compute(), '%')

score = 1-np.abs(np.diag((m[:,:n_sv].T).dot(u[:,:n_sv])))
print('Float64 - left:', score)

score = 1-np.abs(np.diag((nt[:n_sv,:]).dot(vt[:n_sv,:].T)))
print('Float64 - right:', score)

score = 1-np.abs(np.diag((m[:,:n_sv].T).dot(q[:,:n_sv])))
print('Float32 - left:', score)

score = 1-np.abs(np.diag((nt[:n_sv,:]).dot(zt[:n_sv,:].T)))
print('Float32 - right:', score)
```

gives 

```
Relative Frobenius Difference:  24.319256431337795 %
Relative Frobenius Difference Float64:  24.234345384226415 %
Relative Frobenius Difference Float32:  24.546193818628193 %
Float64 - left: [5.45306026e-08 9.93339701e-01 9.51934863e-01 8.52918661e-01
 8.25769860e-01 8.47140783e-01 9.19779391e-01 9.46797727e-01
 8.79011422e-01 9.48384044e-01]
Float64 - right: [9.99200722e-16 8.89398553e-01 9.34052564e-01 6.87783334e-01
 8.04362074e-01 9.60763979e-01 9.34439126e-01 9.58198904e-01
 8.37106715e-01 9.10802639e-01]
Float32 - left: [-1.03100417e-09  9.56865929e-01  9.88275680e-01  8.86883083e-01
  9.17216946e-01  9.49672255e-01  7.34570274e-01  9.83579056e-01
  9.62205157e-01  7.99416893e-01]
Float32 - right: [-7.06722902e-09  6.57628813e-01  7.65107271e-01  8.18695384e-01
  6.54883672e-01  8.22528507e-01  8.42739665e-01  9.78176425e-01
  9.92432486e-01  8.75836096e-01]
```

**Comment by mrocklin** [comment]
This is exciting work @RogerMoens .  Thank you for putting it together.  cc @eric-czech 

For testing we often use pytest and the files in `dask/array/tests/test_linalg.py`.  These are run on every pull request, so if we can put something there it will alert future developers if someone mistakenly reverses this improvement.  Do you have thoughts on what we can do as a test here?  

The experiments that you've run above are probably more than constraining enough.  We probably have to use `assert` statements rather than `print`, and choose values that are very likely to be true, even under fully random conditions.  We probably can't make timing assertions, just because CI servers tend to be under highly variable load, and events like the Python garbage collector can insert arbitrary wait times.

Are you familiar with testing in Python with pytest?  You can run tests from the root directory of the repository by doing the following:

```
pytest dask/array/tests/test_linalg.py
```

You can write tests like the following: https://docs.dask.org/en/latest/develop.html#test

I'm happy to provide more assistance here, including example starter tests, if that would help.

**Comment by RogerMoens** [comment]
@mrocklin it might be a bit harder to test the randomized methodology as one should not really look at left singular and right singular values, but to the reconstructed matrix. So some thought need to go into it to pin that down. If I find the time over the weekend I will have a look on what tests might be viable and constraining enough, taking the CI constraints into account. Thanks for the pytest reference and the dask website reference, I will have a closer look to those! 

Should I make a new PR for the testing or can I just join it in this one, what is usually done?

Cheers,


Roger

File: dask/array/linalg.py [hunk 1]
```diff
@@ -675,9 +675,13 @@ def compression_matrix(data, q, n_power_iter=0, seed=None, compute=False):
         state = seed
     else:
         state = RandomState(seed)
+    if (data.dtype).itemsize <= 4:
```
Comment (eric-czech) [inline dask/array/linalg.py:678]: I think it would probably make more sense if this maintained consistency with numpy linalg dtype preservation (at least for `svd` and `qr`), which would mean anything other than float32, complex64 or complex128 would become float64.

Perhaps:

```suggestion
    datatype = data.type
    if datatype not in {np.float32, np.complex64, np.complex128}:
        datatype = np.float64
```

I don't know why numpy does it that way, but I can see value in dask svd/svd_compressed and numpy svd having the same semantics.

Comment (RogerMoens) [reply dask/array/linalg.py:678]: @eric-czech, if I am right LAPACK does not provide routines for 16 bit or 128 bit calculations (not sure either whether numpy/scipy provides support for those type towards LAPACK). So numpy/scipy automatically promotes it to float32 when it is lower than 32 bits. It keeps complex datatypes, but everything else gets translate into float I believe. The same will be true for the QR decomposition I believe. For complex input data types it is comes naturally to sample its columns following the same precision, but with non-complex weights. Don't see why this doesn't happen for np.complex32. 

It is a good suggestion to maintain consistency with numpy! I will review this on Wednesday (11/11).

Comment (RogerMoens) [reply dask/array/linalg.py:678]: @eric-czech: so reviewing your suggestion. First of all, I couldn't find the suggestion you made in numpy linalg, but in sklearn they have something similar to your suggestion so I presume you meant sklearn. We should keep in mind that we are sampling a matrix columns of the "data" variable, with a certain dtype. We want to make sure that that dtype doesn't get promoted to a higher one. Therefore, I have made a small experiment to see how numpy reacts on different datatypes:

```python
import numpy as np

types = {'int', 'uint', 'float', 'complex'}
bits = {'int': {'8', '16', '32', '64'},
        'uint': {'8', '16', '32', '64'},
        'float': {'32', '64'},
        'complex': {'64', '128'}}

x = np.random.rand(100,10)
for dt in types:
    for bit in bits[dt]:
        u, s, vt = np.linalg.svd(x.astype(dt+bit, copy=False), full_matrices=False)
        print(dt+bit+':', u.dtype, s.dtype, vt.dtype)
        del(u, s, vt)
```

to my surprise this is the outcome (same behaviour for single chunk and multiple chunks dask svd and qr):

> complex64: complex64 float32 complex64
> complex128: complex128 float64 complex128
> int8: float64 float64 float64
> int16: float64 float64 float64
> int32: float64 float64 float64
> int64: float64 float64 float64
> uint8: float64 float64 float64
> uint16: float64 float64 float64
> uint32: float64 float64 float64
> uint64: float64 float64 float64
> float32: float32 float32 float32
> float64: float64 float64 float64

Complex input is what to be expected, float input is what is to be expected, only thing that is remarkable is that all integers types are all converted to float64. I am not sure whether this is a LAPACK thingy or rather a numpy decision. And although the impact might be little for in-memory calculations, blowing up your int8 matrix to float64 might be dramatic if you are not aware of it, already float32 is times 4, if a reconstruction of the matrix is desired of course. 

Nevertheless, if we want to maintain the numpy standard, then everything gets promoted to float64, EXCEPT for complex64 (which is float32 under the hood) and float32 itself. **sklearn makes a mistake by also promoting complex64 via float64**, as the reconstructed matrix will become complex128 to compensate for the float64 singular values:

```python
import numpy as np

u = np.random.rand(100,10).astype('complex64', copy=False)
s_32 = np.diag(np.random.rand(10,10)).astype('float32', copy=False)
s_64 = np.diag(np.random.rand(10,10)).astype('float64', copy=False)
vt = np.random.rand(10,100).astype('complex64', copy=False)

a = (u*s_32)@vt
b = (u*s_64)@vt

print(a.dtype, b.dtype)
```

> complex64 complex128

Since we are randomly sampling, we only want non-complex weights, such that only float32 or float64 can be the dtype of our sampling matrix omega. Simply said we want our datatype to be

**float32** in case input matrix is: **float32 or complex64**
**float64** in case input matrix is: **everything else**.

I will to find out how to make a proposal for this :p:.

Comment (eric-czech) [reply dask/array/linalg.py]: > It keeps complex datatypes, but everything else gets translate into float I believe. The same will be true for the QR decomposition I believe

Nice, good to know.

> First of all, I couldn't find the suggestion you made in numpy linalg, but in sklearn they have something similar to your suggestion so I presume you meant sklearn

I meant numpy.linalg.svd though I wasn't referring to code, just my own loop I ran at some point in the past like you ran checking output dtypes based on input dtypes.  I didn't know that "sklearn makes a mistake by also promoting complex64 via float64" and honestly hadn't thought that much about how complex types should get handled.

> Simply said we want our datatype to be
> float32 in case input matrix is: float32 or complex64
> float64 in case input matrix is: everything else.

Sounds good to me, thanks for digging into it more!

**Comment by RogerMoens** [comment]
I have added a test case, found out that complex numbers are not handled by tsqr and some int types are also not accepted. So I have added two test cases for float32 and float64.

**Review by eric-czech — APPROVED** [review]

**Comment by eric-czech** [comment]
LGTM, perhaps @TomAugspurger could take a pass before merging.

FYI re:

> Should I make a new PR for the testing or can I just join it in this one, what is usually done?

I think it would be very unusual for any dask PR to not include tests.  I have yet to see one like that at least.

**Comment by mrocklin** [comment]
This seems fine to me.  Merging. 

Thank you @RogerMoens for the code contribution and @eric-czech for reviewing!
